src/core/coap/coap.hpp

```diff
@@ -426,11 +426,24 @@ class ResponsesQueue
  * This class implements the CoAP client and server.
  *
  */
-class Coap : public InstanceLocator
+class CoapBase : public InstanceLocator
 {
     friend class ResponsesQueue;
 
 public:
+    /**
+     * This function pointer is called to send a CoAP message.
+     *
+     * @param[in]  aCoapBase     A pointer to the CoAP agent.
+     * @param[in]  aMessage      A reference to the message to send.
+     * @param[in]  aMessageInfo  A reference to the message info associated with @p aMessage.
+     *
+     * @retval OT_ERROR_NONE     Successfully sent CoAP message.
+     * @retval OT_ERROR_NO_BUFS  Failed to allocate retransmission data.
+     *
+     */
+    typedef otError (*SendCallback)(CoapBase *aCoapBase, Message &aMessage, const Ip6::MessageInfo &aMessageInfo);
```

Comment:
How about calling this type `Sender` instead of `SendCallback`?
(I feel it's not a callback and more of a customization point (or method delegation) from base to sub-classes.
`Sender` also matched the `Interceptor` type defined after this type.